tools/clang/lib/SPIRV/DeclResultIdMapper.cpp

```diff
@@ -140,21 +144,65 @@ std::vector<uint32_t> DeclResultIdMapper::collectStageVars() const {
   return vars;
 }
 
+namespace {
+/// A class for managing stage input/output locations to avoid duplicate uses of
+/// the same location.
+class LocationSet {
+public:
+  LocationSet() : usedLocs(64, false), nextLoc(0) {}
+
+  /// Uses the given location.
+  void useLoc(uint32_t loc) { usedLocs.set(loc); }
+
+  /// Uses the next available location.
+  uint32_t useNextLoc() {
+    while (usedLocs[nextLoc])
+      nextLoc++;
+    usedLocs.set(nextLoc);
+    return nextLoc++;
+  }
+
+private:
+  llvm::SmallBitVector usedLocs; ///< All previously used locations
+  uint32_t nextLoc;              ///< Next available location
+};
+} // namespace
+
 void DeclResultIdMapper::finalizeStageIOLocations() {
-  uint32_t nextInputLocation = 0;
-  uint32_t nextOutputLocation = 0;
+  std::vector<const StageVar *> inputVars;
+  std::vector<const StageVar *> outputVars;
+
+  LocationSet inputLocs;
+  LocationSet outputLocs;
 
-  // TODO: sort the variables according to some criteria first, e.g.,
-  // alphabetical order of semantic names.
-  for (auto &var : stageVars) {
+  for (const auto &var : stageVars)
     if (!var.isSpirvBuitin()) {
-      if (var.getSigPoint()->IsInput()) {
-        theBuilder.decorateLocation(var.getSpirvId(), nextInputLocation++);
+      // Only SV_Target, SV_Depth, SV_DepthLessEqual, SV_DepthGreaterEqual,
+      // SV_StencilRef, SV_Coverage are allowed in the pixel shader.
+      // Arbitrary semantics are disallowed in pixel shader.
+      if (var.getSemantic()->GetKind() == hlsl::Semantic::Kind::Target) {
+        theBuilder.decorateLocation(var.getSpirvId(), var.getSemanticIndex());
+        outputLocs.useLoc(var.getSemanticIndex());
+      } else if (var.getSigPoint()->IsInput()) {
+        inputVars.push_back(&var);
       } else if (var.getSigPoint()->IsOutput()) {
-        theBuilder.decorateLocation(var.getSpirvId(), nextOutputLocation++);
+        outputVars.push_back(&var);
       }
     }
-  }
+
+  // Sort stage input/output variables alphabetically
+  const auto comp = [](const StageVar *a, const StageVar *b) {
+    // TODO: error out for the same semantic name and index.
```

Comment:
This seems like a very important error to have sooner rather than later. Can't you add `emitError...` right now?